File: AccuPay.Core.UnitTests/Mothers/LoanMother.cs
using AccuPay.Core.Entities;
using Moq;
using System;
using System.Collections.Generic;
using System.Linq;

namespace AccuPay.Core.UnitTests
{
    public static class LoanMother
    {
        public static Loan Simple()
        {
            return new Loan()
            {
                RowID = It.IsAny<int>(),
                EmployeeID = It.IsAny<int>(),
                OrganizationID = It.IsAny<int>(),
                LoanTypeID = It.IsAny<int>(),
                LoanNumber = It.IsAny<string>(),
                TotalLoanAmount = It.IsAny<decimal>(),
                TotalBalanceLeft = It.IsAny<decimal>(),
                DedEffectiveDateFrom = It.IsAny<DateTime>(),
                DeductionAmount = It.IsAny<decimal>(),
                Status = It.IsAny<string>(),
                DeductionPercentage = It.IsAny<decimal>(),
                DeductionSchedule = It.IsAny<string>(),
                Comments = It.IsAny<string>()
            };
        }

        public static Loan WithLoanAmountAndDeductionAmount(
            decimal totalLoanAmount,
            decimal originalDeductionAmount,
            List<LoanTransaction> loanTransactions = null)
        {
            var loan = Simple();

            loan.TotalLoanAmount = totalLoanAmount;
            loan.TotalBalanceLeft = loan.TotalLoanAmount;
            loan.OriginalDeductionAmount = originalDeductionAmount;
            loan.DeductionAmount = loan.OriginalDeductionAmount;

            loan.LoanTransactions = loanTransactions;
            loan.TotalBalanceLeft -= loanTransactions?.Sum(l => l.PrincipalAmount) ?? 0;

            return loan;
        }

        public static Loan ForYearlyLoanInterest(
            decimal totalLoanAmount,
            decimal deductionPercentage,
            decimal basicMonthlySalary,
            decimal originalDeductionAmount,
            List<LoanTransaction> loanTransactions = null)
        {
            var loan = WithLoanAmountAndDeductionAmount(
                totalLoanAmount: totalLoanAmount,
                originalDeductionAmount: originalDeductionAmount,
                loanTransactions: loanTransactions);

            loan.DeductionPercentage = deductionPercentage;
            loan.BasicMonthlySalary = basicMonthlySalary;

            return loan;
        }
    }
}

File: TestPayrollSystem/Mothers/LoanMother.vb
Option Strict On

Imports AccuPay.Core.Entities
Imports Moq

Public Class LoanMother

    Public Shared Function Simple() As LoanSchedule

        Return New LoanSchedule() With {
            .RowID = It.IsAny(Of Integer),
            .EmployeeID = It.IsAny(Of Integer),
            .OrganizationID = It.IsAny(Of Integer),
            .LoanTypeID = It.IsAny(Of Integer),
            .LoanNumber = It.IsAny(Of String),
            .TotalLoanAmount = It.IsAny(Of Decimal),
            .TotalBalanceLeft = It.IsAny(Of Decimal),
            .DedEffectiveDateFrom = It.IsAny(Of Date),
            .DeductionAmount = It.IsAny(Of Decimal),
            .Status = It.IsAny(Of String),
            .DeductionPercentage = It.IsAny(Of Decimal),
            .DeductionSchedule = It.IsAny(Of String),
            .Comments = It.IsAny(Of String)
        }

    End Function

    Public Shared Function WithLoanAmountAndDeductionAmount(
        totalLoanAmount As Decimal,
        originalDeductionAmount As Decimal,
        Optional loanTransactions As List(Of LoanTransaction) = Nothing) As LoanSchedule

        Dim loan = Simple()

        loan.TotalLoanAmount = totalLoanAmount
        loan.TotalBalanceLeft = loan.TotalLoanAmount
        loan.OriginalDeductionAmount = originalDeductionAmount
        loan.DeductionAmount = loan.OriginalDeductionAmount

        loan.LoanTransactions = loanTransactions
        loan.TotalBalanceLeft -= If(loanTransactions?.Sum(Function(l) l.PrincipalAmount), 0)

        Return loan
    End Function

    Public Shared Function ForYearlyLoanInterest(
        totalLoanAmount As Decimal,
        deductionPercentage As Decimal,
        basicMonthlySalary As Decimal,
        originalDeductionAmount As Decimal,
        Optional loanTransactions As List(Of LoanTransaction) = Nothing) As LoanSchedule

        Dim loan = WithLoanAmountAndDeductionAmount(
            totalLoanAmount:=totalLoanAmount,
            originalDeductionAmount:=originalDeductionAmount,
            loanTransactions:=loanTransactions)

        loan.DeductionPercentage = deductionPercentage
        loan.BasicMonthlySalary = basicMonthlySalary

        Return loan

    End Function

End Class
